---
title: "Capstone"
author: "Sung"
date: "11/13/2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Capstone

## Week 1

### The big question

A fundamental question in human biology is understanding how we develop from a fetus to a grown human being. One way to do this is to study how the human brain changes over time. In a recent study, Jaffe et al. measured gene expression from different individuals across the human lifespan. They were looking for genes that showed patterns of expression that changed over time as people aged. You can find the paper here: http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4281298/. You don’t need to understand every sentence in this paper, but it will be a good idea to spend some time trying to read it.

The primary data generated by this group is transcriptome sequencing data (known as RNA-seq) from human post-mortem brains, sequenced on the Illumina platform. You have learned about this type of data in the Introduction to Genomic Technologies class and learned pieces of the analysis pipeline along the way. The specific question we will ask in the capstone project is:

What is the difference between gene expression in fetal and adult brains?

As always in genomics, this dataset has unique challenges. For example, this is data from human brains and all samples are obtained after the death of the individual (also called post-mortem). It is very possible that brain expression changes after death and that this change is bigger the more time has passed after death (also called the post-mortem interval).

### Objectives

In this project you will be completing the following tasks to answer the fundamental question we have posed:

1. You will read the original research article and begin to understand the fundamental scientific question you are studying.
2. You will download the raw sequence data and the meta-data (phenotypic information and technological information) from a public database, and pre-process the data by aligning it to the genome.
3. You will then perform a quality control on the alignments.
4. You will calculate expression measurements at the gene count level.
5. You will do an exploratory analysis to identify major features of the data and figure out which model to build.
6. You will fit statistical models to identify genes that are different between fetal brains and adult brains.
7. You will integrate you results with data from other projects to identify biological patterns.
8. You will document your analysis in detail to demonstrate reproducibility and turn in a brief writeup.

These steps will be evaluated with a combination of quizzes and peer grading to make sure you are progressing through the capstone. 

### Why re-analysis?

Most datasets are either highly practical and proprietary - think of data collected by Google, Facebook, and Airbnb - or not very practical but freely available - like data from astronomical observatories. One of the most incredible things about the genomic data revolution is that the data are both highly practical for solving biological problems and very often freely available.

Many scientists spend their entire careers cleverly reusing data that has been generated by other people and made freely available on the internet. In this project, we will be focusing on the reanalysis of a publicly available data set. The question we will be answering is a fundamental question in biology that is an area of intense ongoing scientific investigation. The type of reanalysis you are doing is just like the type you would see happen in genomic labs all over the world. 

### Why is it so hard?

We will guide you through the overall process and point you to useful tools, you may have to look up information, download data, or figure out how to put a pipeline together to complete the project. This will likely involve some frustration because there isn’t one right answer or step by step process. This is on purpose. The goal of this capstone project is to demonstrate you can figure out how to execute genomic data analyses on your own. As people who regularly work with and hire genomic data scientists, this independence and ability to solve problems is as important as any of the specific tools or techniques you have learned.

Just like in a real data analysis this doesn’t mean you are on your own. We expect you to complete your projects on your own, but should definitely ask for help and discuss with your classmates. We hope that you will get a lot out of the experience and come out confident in your ability to solve genomic problems with data science.

## Week2

### The dataset

Human brain across the development

This project uses data that generated to explore gene expression changes across human brain development from fetal samples to old age. 

The project started with taking mRNA and RNA sequencing. 

#### What has been done?

Sequenced the polyA transcriptome of 36 human frontal cortex samples at high coverage across 6 important life stages. 6 samples per group. 

Confounding variable balance attempt RNA integrity number (measure of RNA quality), and post-mortem interval. 

#### Data progress steps

take some of sequencing reads -> align to genome (BAM file output) -> quantify expression (output is count) 

### Phenotype information

The data is deposited into the Short Read Archive as BioProject PRJNA245228 (http://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA245228) . This project has 48 different samples, where a sample is a different person and library extraction method. The authors studied 3 different populations of RNA: “total”, “cytosol”, and “nucleus”. These populations refers to RNA present in different parts of the cell, “total” indicates that the RNA comes from both the cytosol and the nucleus. The authors did polyA+ selection to removed ribosomal RNA; this is the standard assay for RNA sequencing.

For “total” RNA samples, the authors studied 6 different age groups: fetal (<0 years), infant (0-1 years), child (1-10 years), adolescent (10-20 years), adult (20-50 years) and old, (50+ years), and each age group had 6 individuals.

The individual samples has names such as “Br5341C1_DLPFC_polyA_RNAseq_cytosol” (http://www.ncbi.nlm.nih.gov/biosample/2999562), which indicates individual “Br5341C1” from the “DLPFC” project (the authors studied the human dorsolateral prefrontal cortex) isolated using “polyA” fromt the “cytosol”. In the link you can see some additional phenotype information such as age (-0.4 years), sex (male), RIN number (10; this is a measure of RNA quality). The age is measured as weeks after conception converted to years before term (which is why the age is negative for this sample; it indicates a fetal sample).

### Getting the data from SRA

The previous section helps us to locate the samples. We now need to get the data out of SRA in the form of a set of FASTQ files.

On the webpage http://www.ncbi.nlm.nih.gov/biosample/2999518, in the top right corner there is a header called “Related Information”, with a link to SRA. Clicking on that link takes us to an SRA page http://www.ncbi.nlm.nih.gov/sra?LinkName=biosample_sra&from_uid=2999518. Here we see important information. First, internally in SRA this BioSample is called SRX683793; two Runs with ids SRR1554535 and SRR2071346 are associated with this sample. The ids beginning with SRX are called experiment ids and the ones beginning with SRR are called run ids. In this case it means that this particular sample was sequenced twice. Each sequencing run will give us a FASTQ file, and we will therefore have two FASTQ files associated with this sample.

Side note: internally in SRA all data is stored in a special SRA format, which is - frankly - irritating to deal with. But it allows us to potentially retrieve the data in FASTA, FASTQ and SRA formats. We want to get the FASTQ files.

Clicking on either of the Run ids at the bottom of the page takes us to http://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1554535, a screen shot is below.

Confusingly, the page you land at is labelled “download” but you can only download a pileup file. Instead click at the top, at the other (!) download tab. Here you get taken to a page where you can only enter experiment ids, beginning with SRX. Doing this, lands you on a page like this http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?exp=SRX683792&cmd=search&m=downloads&s=seq, which allows you to download FASTQs.

An alternative to the web interface is to use the sra toolkit, a command line utility you can find at http://trace.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software. There are several command line utilities for accessing SRA. A classic is the fastq-dump command; an example of using it is

fastq-dump -v --gzip SRR1554534

This will output a gzipped FASTQ file. The utility only appears to support run ids, not experiment ids.

### Which samples to use

As stated elsewhere, we will be doing a comparison between fetal and adult samples. But the samples themselves are quite big. The comparison will of course be better if you use the full 6 vs. 6 samples, but you can select down to only 3 vs. 3 and still get full credit.

The full list of fetal and adult samples are as follows:
Fetal

R3452_DLPFC_polyA_RNAseq_total; SRX683795; SRR1554537 , SRR2071348

(http://www.ncbi.nlm.nih.gov/biosample/2999520)

R3462_DLPFC_polyA_RNAseq_total; SRX683796; SRR1554538, SRR2071349

(http://www.ncbi.nlm.nih.gov/biosample/2999521)

R3485_DLPFC_polyA_RNAseq_total; SRX683799; SRR1554541, SRR2071352

(http://www.ncbi.nlm.nih.gov/biosample/2999524)

R4706_DLPFC_polyA_RNAseq_total; SRX683824; SRR1554566, SRR2071377

(http://www.ncbi.nlm.nih.gov/biosample/2999549)

R4707_DLPFC_polyA_RNAseq_total; SRX683825; SRR1554567, SRR2071378

(http://www.ncbi.nlm.nih.gov/biosample/2999550)

R4708_DLPFC_polyA_RNAseq_total; SRX683826; SRR1554568, SRR2071379

(http://www.ncbi.nlm.nih.gov/biosample/2999551)
Adults

R2869_DLPFC_polyA_RNAseq_total; SRX683793; SRR1554535, SRR2071346

(http://www.ncbi.nlm.nih.gov/biosample/2999518)

R3098_DLPFC_polyA_RNAseq_total; SRX683794; SRR1554536, SRR2071347

(http://www.ncbi.nlm.nih.gov/biosample/2999519)

R3467_DLPFC_polyA_RNAseq_total; SRX683797; SRR1554539, SRR2071350

(http://www.ncbi.nlm.nih.gov/biosample/2999522)

R3969_DLPFC_polyA_RNAseq_total; SRX683814; SRR1554556, SRR2071367

(http://www.ncbi.nlm.nih.gov/biosample/2999539)

R4166_DLPFC_polyA_RNAseq_total; SRX683819; SRR1554561, SRR2071372

(http://www.ncbi.nlm.nih.gov/biosample/2999544)

R2857 DLPFC polyA+ transcriptome; SRX683792; SRR1554534, SRR2071345

(http://www.ncbi.nlm.nih.gov/biosample/2731373)

## Week 3

### Task1 instruction

The purpose of genomic data science is to answer fundamental questions in biology. Before starting on the data analysis process, the first step is always to understand the scientific question you are trying to answer. The first assessment in the capstone is to make sure you have read and understood the background on the question we are trying to answer. Please read the following paper: http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4281298 and then answer the questions in the Task 1 Quiz. 

### The Results

Introduction

- Transcriptome analysis of human brain provide fundamental insight into development and disease.

Method

- RNA sequencing from dorsolateral prefrontal cortex in 36 samples across the life span.
- Applied approach called DER Finder (Differential Expressed Region Finder) to perform differential expression analysis at single base resolution. 
- Permutation test to assess statistical significance of the identified differentially expressed regions. 
- DER, ranked by significance, annotated to known gene annotation. 

Result

- 40% of differentially expressed regions were annotated to intronic or intergenic sequence. Three infant and three adult samples were used to validate by separating RNA into nucleic and cytosolic fraction. Sequences were observed in both condition suggesting nuclear export. 
- Genome wide association study to understand clinical importance. GWAS was compared with known gene expression of multiple dieseases. Etiology overlapped with development phases. Negative controlw as Type 2 Diabetes. 
- Took 50k regions to looked in other projects. 
- Many of differentially expressed regions are conserved in the mouse. 
- Among other human samples outside of the brain, found evidence of expression and also regulation. Fetal samples are similar to stem cells.
- Fetal brain has the largest fraction of genome expressed by tabulating contiguous reigon expressed over particular threshold. 